File: Div2A/checkBalancedParentheses.py
import re

str = input()

str = re.sub("[^()\\[\\]\\{\\}]", '', str)

def checkBalancedParentheses(str):
    stack = []
    for i in str:
        if (stack):
            if stack[-1] == '[' and i == ']' or stack[-1] == '(' and i == ')' or stack[-1] == '{' and i == '}':
                stack.pop()
        else:
            stack.append(i)
    
    if not stack:
        return "balanced"
    return "not balanced"

print(checkBalancedParentheses(str))

File: Div2A/checkBalancedParentheses.cpp
#include <iostream>
#include <stack>
#include <regex>
using namespace std;
bool checkParentheses(string str) {
    stack<char> S;
    for (int i = 0; i < str.length(); i++)
    {
        if(S.empty()) {
            S.push(str[i]);
        } else {
            if (S.top() == '[' && str[i] == ']' || S.top() == '(' && str[i] == ')' || S.top() == '{' && str[i] == '}')
                S.pop();
            else
                S.push(str[i]);
        }
    }
    if (S.empty())
        return true;
    return false;
}

int main() {
    string str;
    cin>>str;
    regex r("[^()\\[\\]\\{\\}]*");
    str = regex_replace(str,r,"");
    cout<<checkParentheses(str);
}